### ISSUE TITLE
torch.multinomial is misnamed. 

### ISSUE THREAD

**Issue by odysseus9672** [pytorch/pytorch#36036]
The [multinomial distribution](https://en.wikipedia.org/wiki/Multinomial_distribution) has two inputs: weights $\mathbf{p}$ and the number of trials $N$, and one output the number of results in each category $\mathbf{n}$. Note that the concept of with or without replacement doesn't enter the picture, and the output as the same shape as the weights, but with a different data type (an integer that sums to $n$ along the axis that the weights sum to 1). 

There is a distribution that related to the multinomial distribution by the question of replacement, and that's the [multi-variate hypergeometric distribution](https://en.wikipedia.org/wiki/Hypergeometric_distribution#Multivariate_hypergeometric_distribution), but that, too, will have an output with the same shape as weights (and weights will have to contain integers, instead of floats). 

The hypergeometric distribution is about drawing a fixed number of objects from a population that has a fixed number of each type of object without replacing between individual draw, and then counting how many of each type you drew. A multinomial distribution can be gotten from the same process where the objects are replaced after each draw.

Regardless, that's not what torch.multinomial is returning. When you're returning the result of each draw, you're repeatedly sampling either the [categorical distribution](https://en.wikipedia.org/wiki/Categorical_distribution) (the with replacement variant), or a distribution I cannot find a name for (draw?). Although the constraint that the number of draws be less than the number of categories means that the without replacement variant doesn't correspond to any probability distribution I know of, at all.

So, I would suggest renaming torch.multinomial to torch.categorical.

**Comment by zou3519** [comment]
cc @vincentqb what do you think?

**Comment by vincentqb** [comment]
When reading the [documentation](https://pytorch.org/docs/stable/torch.html#torch.multinomial), I see 
```
Returns a tensor where each row contains num_samples indices sampled from the multinomial probability distribution located in the corresponding row of tensor input.
```
Are you saying this is not correct?

The categorical distribution is a special case of the multinomial distribution, in that it gives the probabilities of potential outcomes of a single drawing rather than multiple drawings (see [wiki](https://en.wikipedia.org/wiki/Categorical_distribution)), and in that sense there is no with/without replacement for categorical distribution. In particular, 

* the categorical distribution in that sense is `torch.multinomial(input, num_samples=1, replacement=True)`. Note also that we already have `torch.distributions.categorical`, see [here](https://pytorch.org/docs/stable/distributions.html#torch.distributions.categorical.Categorical), and you can see that the code is doing exactly this when invoking `torch.multinomial`.
* the multinomial distribution is `torch.multinomial(input, num_samples, replacement=True)`
* the multivariate hypergeometric is `torch.multinomial(input, num_samples, replacement=False)`

Is that what you meant?

**Comment by alicanb** [comment]
We actually talked about this a while ago: https://github.com/probtorch/pytorch/issues/46 and @gchanan mentions a similar request here https://github.com/pytorch/pytorch/issues/33122#issuecomment-594151376 

The only argument against this afaik is that it's a big breaking change.

**Comment by odysseus9672** [comment]
> Are you saying this is not correct?

Yes. A sample from the multinomial probability distribution would be the histogram of the number of times each index appears in the return given when `replacement=True`. 

You also raise a good point. The categorical distribution would, likewise, be a vector like `[0, 0, 1, 0]` if the function as currently implemented returned `[2]`. So the suggestion to rename it `torch.categorical` is bad. 

The behavior that this seems to most mimic is the relationship between a <A HREF="https://en.wikipedia.org/wiki/Poisson_point_process">Poisson process</A>, the output of which is a set of points at particular places on a real number line, and a <A HREF="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson distribution</A> which describes the distribution of the number of points in a fixed interval of fixed size generated by a Poisson process. It's not in Wikipedia, but what you're sampling is a <A HREF="https://people.cs.kuleuven.be/~raf.vandebril/homepage/publications/phd/node38.html">multinomial process</A>. 

So, I would suggest, at minimum, changing the documentation to read:

> Returns a tensor where each row contains num_samples indices sampled from a multinomial process located in the corresponding row of tensor input.
and similar for `torch.categorical`.

Ideally, I think that the function `torch.multinomial` should be split into two functions with signatures like
```
torch.multinomial_proc(input, num_samples, *, generator=None, out=None) → LongTensor
torch.hypergeometric_proc(input, num_samples, *, generator=None, out=None) → LongTensor
``` 
to make it unambiguous. Since I don't know of any urgent demand to sample the actual multinomial distribution from within pytorch, this change should be something that can be done gradually with deprecation.
